job-server/src/spark.jobserver/AkkaClusterSupervisorActor.scala

```diff
@@ -0,0 +1,217 @@
+
+      var contextName = ""
+      do {
+        contextName = java.util.UUID.randomUUID().toString().substring(0, 8) + "-" + classPath
+      } while (contexts contains contextName)
+
+      startContext(contextName, mergedConfig, true) { ref =>
+        originator ! (contexts(contextName), resultActors(contextName))
+      } { err =>
+        originator ! ContextInitError(err)
+      }
+
+    case GetResultActor(name) =>
+      sender ! resultActors.get(name).getOrElse(globalResultActor)
+
+    case GetContext(name) =>
+      if (contexts contains name) {
+        sender ! (contexts(name), resultActors(name))
+      } else {
+        sender ! NoSuchContext
+      }
+
+    case StopContext(name) =>
+      if (contexts contains name) {
+        logger.info("Shutting down context {}", name)
+
+        context.watch(contexts(name))
+        contexts(name) ! PoisonPill
+        resultActors.remove(name)
+        sender ! ContextStopped
+      } else {
+        sender ! NoSuchContext
+      }
+
+    case Terminated(actorRef) =>
+      val name: String = actorRef.path.name
+      logger.info("Actor terminated: {}", name)
+      contexts.foreach { kv => if (kv._2 == actorRef) contexts.remove(kv._1) }
+      resultActors.foreach { kv => if (kv._2 == actorRef) resultActors.remove(kv._1) }
+  }
+
+  private def initContext(actorName: String, ref: ActorRef, timeoutSecs: Long = 1)
+                         (ctxName: String, ctxConf: Config, isAdHoc: Boolean)
+                         (successFunc: ActorRef => Unit, failureFunc: Throwable => Unit): Unit = {
+    import akka.pattern.ask
+
+    val resultActor = if (isAdHoc) globalResultActor else context.actorOf(Props(classOf[JobResultActor]))
+    (ref ? JobManagerActor.Initialize(
+      daoActor, Some(resultActor), ctxName, ctxConf, isAdHoc, self))(Timeout(timeoutSecs.second)).onComplete {
+      case Failure(e:Exception) =>
+        logger.info("Failed to send initialize message to context " + ref, e)
+        ref ! PoisonPill
+        failureFunc(e)
+      case Success(JobManagerActor.InitError(t)) =>
+        logger.info("Failed to initialize context " + ref, t)
+        ref ! PoisonPill
+        failureFunc(t)
+      case Success(JobManagerActor.Initialized(_)) =>
+        logger.info("SparkContext {} joined", ctxName)
+        contexts(ctxName) = ref
+        resultActors(ctxName) = resultActor
+        successFunc(ref)
+    }
+
+  }
+
+  private def startContext(name: String, contextConfig: Config, isAdHoc: Boolean)
+                          (successFunc: ActorRef => Unit)(failureFunc: Throwable => Unit) = {
+    require(!(contexts contains name), "There is already a context named " + name)
+    val contextActorName = "jobManager-" + java.util.UUID.randomUUID().toString.substring(16)
+
+    logger.info("Starting context with actor name {} ", contextActorName)
+
+    val pb = Process(s"$managerStartCommand $contextActorName ${selfAddress.toString}")
```

Comment:
I did this way because it lets me identify when a new actor in the cluster is a JobManagerActor [here](https://github.com/spark-jobserver/spark-jobserver/blob/feature/context-per-jvm/job-server/src/spark.jobserver/AkkaClusterSupervisorActor.scala#L68) by checking the name prefix. I guess this is not strictly required right now, but in a scenario with multiple masters, or any other kind of actor, we can't assume everyone who responds to the identify is a JobManagerActor

Theres probably a better way of doing that check though - if  there is one then yeah having the same names would be clearer.

If we had multiple masters, contexts could be named the same but just prefixing the master address would be enough to get unique names there i think?